includes/wc-user-functions.php

```diff
@@ -120,6 +108,66 @@ function wc_create_new_customer( $email, $username = '', $password = '', $args =
 	}
 }
 
+/**
+ * Create a unique username for a new customer.
+ *
+ * @since 3.6.0
+ * @param string $email New customer email address.
+ * @param array  $new_user_args Array of new user args, maybe including first and last names.
+ * @param int    $suffix Append number to username to make it more unique.
+ * @return string Generated username.
+ */
+function wc_create_new_customer_username( $email, $new_user_args, $suffix = 0 ) {
+	$username_parts = array();
+
+	if ( isset( $new_user_args['first_name'] ) ) {
+		$username_parts[] = sanitize_user( $new_user_args['first_name'], true );
+	}
+
+	if ( isset( $new_user_args['last_name'] ) ) {
+		$username_parts[] = sanitize_user( $new_user_args['last_name'], true );
+	}
+
+	// Remove empty parts.
+	$username_parts = array_filter( $username_parts );
+
+	// If there are no parts, e.g. name had unicode chars, or was not provided, fallback to email.
+	if ( empty( $username_parts ) ) {
+		$email_parts    = explode( '@', $email );
+		$email_username = $email_parts[0];
+
+		// Exclude common prefixes.
+		if ( in_array(
+			$email_username,
+			array(
+				'sales',
+				'hello',
+				'mail',
+				'contact',
+				'info',
+			),
+			true
+		) ) {
+			// Get the domain part.
+			$email_username = $email_parts[1];
+		}
+
+		$username_parts[] = sanitize_user( $email_username, true );
+	}
+
+	$username = wc_strtolower( implode( '', $username_parts ) );
+
+	if ( $suffix ) {
+		$username .= $suffix;
+	}
+
+	if ( username_exists( $username ) ) {
+		return wc_create_new_customer_username( $email, $new_user_args, ++$suffix );
```

Comment:
@mikejolley If the recursion is too much for the stack to handle, PHP will segfault. I don't have any information upon how many levels typically you have to go to for this to occur. If the XDebug PHP extension is installed, then by default xdebug will throw a fatal error at 100. I don't have any data on common names seen on really big stores to suggest how likely that is.